swfiles/+ndbase/estimate_hessian.m

```diff
@@ -0,0 +1,174 @@
+%     expensive.
+%
+% `ivary` (optional)
+% : indices of parameters which were varied in the fit - the final hessian
+%   will be a sqare matrix of size [numel(ivary) x numel(ivary)].
+%   If not provided all parameters will be varied.
+%
+% `niter` (optional)
+% : Number of iterations in optimising step size (step doubled in each
+%   iteration from minimum value of sqrt(eps))
+%
+% `cost_tol` (optional)
+% : Minimum difference in cost function value for a step size to be
+%   considered appropriate. Step size optimisation will be terminated once
+%   this condition is met.
+%
+% ### Output Arguments
+% 
+% `hessian` 
+% : Hessian matrix from which covariance matrix can be calculated.
+%
+% ### Examples
+%```
+% >> % make data
+% >> rng default;
+% >> fit_fun = @(x, params) params(1)*x + params(2)./x + params(3);
+% >> pars = [5,4,3];
+% >> x = linspace(0.1,2);
+% >> y = fit_fun(x, pars);
+% >> e = sqrt(y);
+% >> y = y + e.*randn(size(y));
+% >>
+% >> % fit data
+% >> fcost = @(params) sum(((y-fit_fun(x, params))./e).^2);  % chisq
+% >> [pars_fit, chisq, ~] = ndbase.simplex([],fcost,pars);
+% >>
+% >> % estimate errors on fit parameters
+% >> hess = ndbase.estimate_hessian(fcost, pars_fit);
+% >> cov = inv(hess) * 2.0 * chisq/ (numel(y) - numel(pars_fit));
+% >> perr = sqrt(diag(cov)); % errors on best fit parameters
+%'''
+%
+    % defaults
+    optimise_step = true;
+    min_step = sqrt(eps);
+    step_size = params.*min_step;
+    npar = numel(params);
+    ivary = 1:npar;
+    max_niter = 16; % default number of iterations in optimising step size
+    cost_tol = min_step; % used in step size automation
+    % parse parameters
+    for ikey = 1:2:numel(varargin)
+        switch varargin{ikey}
+            case 'step'
+                optimise_step=false;
+                step_size = varargin{ikey+1};
+                if numel(step_size)==1
+                    % interpret as fractional step size
+                    step_size = params.*step_size;
+                end
+            case 'ivary'
+                ivary = varargin{ikey+1};
+                npar = numel(ivary);
+            case 'niter'
+                max_niter = varargin{ikey+1};
+            case 'cost_tol'
+                cost_tol = varargin{ikey+1};
+        end
+    end
+
+    if numel(step_size) ~= numel(params)
+        error("ndbase:estimate_hessian", ...
+              "If step_size is provided it must be a scalar or vector " + ...
+              "with same number of elements as params");
+    end
+    % enforce minimum step size
+    step_size(abs(step_size) < min_step) = min_step;
+    
+    % calculate jacobian at param
+    hessian = zeros(npar);
+    initial_cost = fcost(params);
+    jac_one_step = zeros(1, npar);
+    cost_one_step = zeros(1, npar);
+    for irow = 1:npar
+        ipar = ivary(irow);
+        success = false;
+        for niter = 1:max_niter
+            params(ipar) = params(ipar) + step_size(ipar);
```

Comment:
I don't think the negative step-size matters if you are at the local minimum - I have a unit test `test_absolute_step_size_negative` that shows this at least for one case. In general though it could depending on the form of the error surface so I thought I might as well give the option. Happy to take `abs` value.